### DISCUSSION TITLE
Empty demes in 2D stepping stone models

### DISCUSSION THREAD

**Discussion by connor-french** [tskit-dev/msprime#2251] [Q&A] [answered]
Howdy y'all! I'd love some feedback on a pattern I'm noticing when running a large 2D stepping stone model (118 x 100 demes). I have a hunch about why this is occurring, but I'm curious what others think.

I am simulating a variety of deme sizes across the "landscape", ranging from N = 6 to N = 100 with moderate migration rates (~0.001 but scaled based on the size of the donor vs receiving population). The deme sizes change at 1000 generation intervals for 22000 generations according to projected responses to environmental change. Before 22000 generations, the demes collapse into two randomly mating populations that correspond with empirical population structure inferred to be more ancient than 22000 generations. I sample 45 individuals from 28 demes spread across the landscape, distributed like as follows:  
![image](https://github.com/tskit-dev/msprime/assets/37662997/069b2f7a-5dd9-46de-8576-2f958eea537f)

The minimum size any sampled deme contracts reaches is 24, while the max size is 97. 

**The problem**
When I conduct simulations using the standard coalescent, a seemingly random subset of ~1/4 of the sampled demes return zero diversity, calculated either directly on the tree sequence or using a genotype matrix of overlaid mutations. When I conduct simulations using a multiple merger coalescent (Dirac) for the early "spatial" phase and the standard coalescent for the later "non-spatial" phase, this problem does not occur.  

What I think is happening is the demographic changes are occurring rapidly and over short time periods on small demes, resulting in a high probability of a polytomy occurring, where multiple descendants must coalesce to a single ancestor. This results in a situation the standard coalescent cannot accommodate. What I find confusing is that the diversities are zero, rather than a value that doesn't correspond with some theoretical expectation. When viewing a tree sequence, the offending samples coalesce recently, but they still coalesce and don't result in infinite waiting times or some other offending process. E.g. in the following tree sequence the two samples corresponding to "pop_25_71" coalesce 100 generations ago, which should correspond to at least some diversity I would think. However, they result in zero diversity (they're towards the left side of the tree seq).
![Screenshot 2024-01-03 at 1 40 36 PM](https://github.com/tskit-dev/msprime/assets/37662997/ddd7cb6f-7dbe-42ef-8aa8-3a40a4103301)

I feel like I'm missing something big, but I can't put my finger on it. My first attempts with smaller landscapes/simpler demographic scenarios don't reproduce the problem, but I can put in more time to create a minimal reproducible example if it's not immediately clear. Thanks so much for any feedback!

**Comment by petrelharp** [comment, marked as answer]
Hm - this is confusing. But let's see - from your image, there is a difference between the samples in the tree, so if you compute diversity with `mode="branch"`, then you should get non-zero diversity (right?). If so, then does zero diversity just mean that those samples happen to not have any mutations on the branches betwen them? You can check this by counting number of mutations using `ts.diversity(..., mode='site', span_normalise=False)`.

Said another way: the total number of mutations separating two samples is Poisson with mean equal to  `mut_rate * ts.diversity(..., mode='branch', span_normalise=False)`; if that is not a large number, then there's some reasonable chance the number of mutations is zero, and the (site) diversity will be zero.

Does this agree with what you're seeing?

**Comment by connor-french** [comment]
Ah, I was forgetting to set mode="branch" on the tree sequence diversity measure! I was expecting that behavior without setting the option.

When I do it, I get a reasonable number for one of them (2522). And I do get a low expected number of mutations (2.5ish), so this makes sense. I'm calculating the per-individual pi in scikit-allel using additional filtering steps to match empirical data as well, so there is a good chance that these get washed out. I guess when using the multiple merger coalescent, G = N^2, so the more recent coalescent times get pushed back at low population sizes, increasing the numebr of mutations. Thanks a ton!

**Reply by petrelharp** [reply]
phew, I'm glad it was that straightforward!
